scipy/spatial/src/distance_pybind.cpp

```diff
@@ -0,0 +1,593 @@
+    });
+    return out;
+}
+
+template <typename Func>
+py::array cdist(const py::object& out_obj, const py::object& x_obj,
+                const py::object& y_obj, const py::object& w_obj, Func&& f) {
+    auto x = npy_asarray(x_obj);
+    auto y = npy_asarray(y_obj);
+    if (x.ndim() != 2) {
+        throw std::invalid_argument("XA must be a 2-dimensional array.");
+    }
+    if (y.ndim() != 2) {
+        throw std::invalid_argument("XB must be a 2-dimensional array.");
+    }
+    const int64_t m = x.shape(1);
+    if (m != y.shape(1)) {
+        throw std::invalid_argument(
+            "XA and XB must have the same number of columns "
+            "(i.e. feature dimension).");
+    }
+
+    std::array<int64_t, 2> out_shape{{x.shape(0), y.shape(0)}};
+    if (w_obj.is_none()) {
+        auto dtype = promote_type_real(common_type(x.dtype(), y.dtype()));
+        auto out = prepare_out_argument(out_obj, dtype, out_shape);
+        DISPATCH_DTYPE(dtype, [&]{
+            cdist_unweighted<scalar_t>(out, x, y, f);
+        });
+        return out;
+    }
+
+    auto w = prepare_single_weight(w_obj, m);
+    auto dtype = promote_type_real(
+        common_type(x.dtype(), y.dtype(), w.dtype()));
+    auto out = prepare_out_argument(out_obj, dtype, out_shape);
+    DISPATCH_DTYPE(dtype, [&]{
+        cdist_weighted<scalar_t>(out, x, y, w, f);
+    });
+    return out;
+}
+
+PYBIND11_MODULE(_distance_pybind, m) {
+    if (_import_array() != 0) {
+        throw py::error_already_set();
+    }
+    using namespace pybind11::literals;
+    m.def("pdist_chebyshev",
+          [](py::object x, py::object w, py::object out) {
+              return pdist(out, x, w, ChebyshevDistance{});
+          },
+          "x"_a, "w"_a=py::none(), "out"_a=py::none());
+    m.def("pdist_cityblock",
+          [](py::object x, py::object w, py::object out) {
+              return pdist(out, x, w, CityBlockDistance{});
+          },
+          "x"_a, "w"_a=py::none(), "out"_a=py::none());
+    m.def("pdist_euclidean",
+          [](py::object x, py::object w, py::object out) {
+              return pdist(out, x, w, EuclideanDistance{});
+          },
+          "x"_a, "w"_a=py::none(), "out"_a=py::none());
+    m.def("pdist_minkowski",
+          [](py::object x, py::object w, py::object out, double p) {
+              if (p == 1.0) {
+                  return pdist(out, x, w, CityBlockDistance{});
+              } else if (p == 2.0) {
+                  return pdist(out, x, w, EuclideanDistance{});
+              } else if (std::isinf(p)) {
+                  return pdist(out, x, w, ChebyshevDistance{});
+              } else {
+                  return pdist(out, x, w, MinkowskiDistance{p});
+              }
+          },
+          "x"_a, "w"_a=py::none(), "out"_a=py::none(), "p"_a=2.0);
+    m.def("cdist_chebyshev",
+          [](py::object x, py::object y, py::object w, py::object out) {
+              return cdist(out, x, y, w, ChebyshevDistance{});
```

Comment:
Continuing FunctionRef:
It appears here, by using an empty instance, the only important information is the type. Thus this would work?
```
cdist<ChebyshevMetric>(out, x, y, w)
```

To accommodate MinkowskiDistance, perhaps also support additional constructor arguments to the type:
```
cdist<MinKowskiMetric>(out, x, w, DistanceMetricInitializer{2, })
```
My proposal is very similar to the template based dispatch mechanism in cktree/cpp.

Reply:
Template deduction works fine. The problem `FunctionRef` solves isn't syntax, it's binary size and compile times.